mender-update/main.cpp

```diff
@@ -0,0 +1,27 @@
+#include <iostream>
+#include <memory>
+
+using namespace std;
+
+
+#include <common/json/json.hpp>
+
+// This method only knows about the interface Json
+void hello_world(std::shared_ptr<json::Json> j)
+{
+    j->hello_world();
+}
+
+
+#include <common/json/impl/boost/boost_json.hpp>
+
+int main()
+{
+    // It is here that we make an object from a concrete type, BoostJson.
+    shared_ptr<json::BoostJson> j = make_shared<json::BoostJson>();
```

Comment:
How will this work with multiple implementations? My assumption was that we would have multiple different implementations for a class and **compilation+linking** will make sure to use the proper implementation on the given platform. So this code would not have to care about the actual implementation at all.

Reply:
This is a bit down the road and I wanted to keep that "road" open. Specially because I will not be deep involved in these design decisions.

FWIW, I was rather envisioning an Abstract Factory pattern, where the client code only this this to get the different objects, and then have a concrete factory, customized at build time (compilation + linking) for the target platform.

Again, this is a bit down the road. We should study how this is done in other projects :nerd_face:

Reply:
> OTOH, maybe I'm just moving the problem around. slightly_smiling_face What do you think? Other ideas?

It's definitely an improvement because it moves the `#ifdef` as close as possible to the place where the actual decision has to be made. My idea, based on Lluis' point and my "C approach" was to have factory functions on top of costructors. So in the above case it would be

```
JsonClass &get_json();
```
where `get_json()` would have multiple implementations using various classes sub-classed from `JsonClass` and the compilation/build process would decide which one to use. Unlike the classes themselves, the function prototype would be just one for all the possible implementations. And the individual (sub-)classes would then have arbitrary members and only the factory function implementations would have to care about the specifics.

Reply:
I think it's not possible to put instances on the stack with that method, is it? It would require using a smart pointer. So `myFunc` from above:

```
int MyFunc() {
    JsonClass json;
    json.DoStuff();
}
```

would not be possible. Instead it would have to be:

```
int MyFunc() {
    unique_ptr<JsonClass> json(get_json());
    json->DoStuff();
}
```

Reply:
Yes, exactly. I figured that out while thinking about this over the weekend. I guess `get_json()` could return `unique_ptr<JsonClass>` directly? Anyway, it's stack allocation replaced by heap allocation which could (would) be a major performance penalty, I guess.

Reply:
>  could (would) be a major performance penalty, I guess.

Only thing ever really hampering us is I/O and write times in terms of performance.

Reply:
Yeah, it's mainly practicality I'm worried about here, not performance. I think the only CPU bound performance areas we are worried about are decompression, and to a much lesser degree, encryption. Everything else is miniscule in comparison.

Reply:
> I took a look at Zephyr, but it's not a good model because it's all C. So I took a look at Qt instead. They seem to use the "one class, separate implementations" approach (see [header](https://github.com/qt/qtbase/blob/1bc7e9e77b2d6b03b995f376b622d4c8c97dd51c/src/corelib/io/qfilesystemengine_p.h) and [implementation](https://github.com/qt/qtbase/blob/1bc7e9e77b2d6b03b995f376b622d4c8c97dd51c/src/corelib/io/qfilesystemengine_unix.cpp)). This means that their header file is riddled with ifdefs though.
> 
> I have a suggestion for an alternative approach:
> 
>     * A shared abstract class (interface).
> 
>     * Separate platform classes, with different names, which inherit from the interface.
> 
>     * Some common platform file somewhere which typedefs a common class name to the specific implementation class name.
> 
> 
> Here's an example:
> 
> ```
> common/json.hpp:
> class AbstractJsonClass {...};
> 
> common/platform.hpp:
> #ifdef POSIX
> typedef PosixJsonClass JsonClass;
> #endif
> 
> common/impl/posix_json.hpp:
> class PosixJsonClass : public AbstractJsonClass {...}
> 
> Other source files:
> int MyFunc() {
>     JsonClass json;
>     json.DoStuff();
> }
> ```
> 
> That would keep the naming complexity in one (or at most a few) files, and would also leave platform specific members fully contained in their own files, instead of spread around everywhere.
> 
> OTOH, maybe I'm just moving the problem around. slightly_smiling_face What do you think? Other ideas?

@kacf now that I have played around with this a bit, I would have to say that I favour the `QT` approach over what we are thinking about doing.

Could we have a small discussion around this during stdup tomorrow?

Reply:
Decision was to use the [Qt approach](https://github.com/qt/qtbase/blob/1bc7e9e77b2d6b03b995f376b622d4c8c97dd51c/src/corelib/io/qfilesystemengine_p.h), but we should avoid having platform-specific **public** class members.